enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/pipes/CreateNodeRegisterPipe.scala

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2002-2017 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher.internal.compatibility.v3_3.runtime.interpreted.pipes
+
+import org.neo4j.collection.primitive.PrimitiveLongCollections
+import org.neo4j.cypher.internal.compatibility.v3_3.runtime.commands.expressions.Expression
+import org.neo4j.cypher.internal.compatibility.v3_3.runtime.helpers.{IsMap, PrimitiveLongHelper}
+import org.neo4j.cypher.internal.compatibility.v3_3.runtime.interpreted.PrimitiveExecutionContext
+import org.neo4j.cypher.internal.compatibility.v3_3.runtime.mutation.makeValueNeoSafe
+import org.neo4j.cypher.internal.compatibility.v3_3.runtime.pipes.{LazyLabel, Pipe, QueryState}
+import org.neo4j.cypher.internal.compatibility.v3_3.runtime.{ExecutionContext, PipelineInformation}
+import org.neo4j.cypher.internal.compiler.v3_3.planDescription.Id
+import org.neo4j.cypher.internal.frontend.v3_3.CypherTypeException
+import org.neo4j.cypher.internal.spi.v3_3.QueryContext
+import org.neo4j.graphdb.{Node, Relationship}
+
+import scala.collection.Map
+
+case class CreateNodeRegisterPipe(ident: String, pipelineInformation: PipelineInformation,
+                                  labels: Seq[LazyLabel], properties: Option[Expression])
+                                   (val id: Id = new Id) extends Pipe {
+
+  private val offset = pipelineInformation.getLongOffsetFor(ident)
+
+  protected def internalCreateResults(state: QueryState): Iterator[ExecutionContext] = {
+    val node: Node = state.query.createNode()
+    val context = PrimitiveExecutionContext(pipelineInformation)
+    val nodeId: Long = node.getId
+    setProperties(context, state, nodeId)
+    setLabels(context, state, nodeId)
+
+    state.copyArgumentStateTo(context)
+    context.setLongAt(offset, nodeId)
+    PrimitiveLongHelper.map(PrimitiveLongCollections.singleton(nodeId), nodeId => context)
+  }
+
+  private def setProperties(context: ExecutionContext, state: QueryState, nodeId: Long) = {
+    properties.foreach { (expr) =>
+      expr(context)(state) match {
```

Comment:
Seems like a great idea, but perhaps done as a second round of improvements, as an optimization to this work.